File: Python/minDistanceDelete.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: minDistanceDelete.py
@Function: 两个字符串的删除操作  动态规划
@Link: https://leetcode-cn.com/problems/delete-operation-for-two-strings/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-16
"""


class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        # method4
        M, N = len(word1), len(word2)
        dp = [0] * (N+1)
        for i in range(1, M+1):
            temp1 = dp[0]
            for j in range(1, N+1):
                temp2 = dp[j]
                if word1[i-1] == word2[j-1]:
                    dp[j] = temp1 + 1
                else:
                    dp[j] = max(dp[j], dp[j-1])
                temp1 = temp2
        return M + N - 2 * dp[-1]


# -------------------------
if __name__ == "__main__":
    n = 2

    solution = Solution()
    max_mul = solution.integerBreak(n)
    print(f"The solution of this problem is {max_mul}")

File: C++/minDistanceDelete.cpp
/**
 * @File    : minDistanceDelete.cpp
 * @Brief   : 两个字符串的删除操作 动态规划
 * @Link    : https://leetcode-cn.com/problems/delete-operation-for-two-strings/
 * @Author  : Wei Li
 * @Date    : 2021-07-16
*/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

class Solution {
public:
    int minDistance(std::string word1, std::string word2) {
        std::vector<std::vector<int>> dp(word1.size() + 1, std::vector<int>(word2.size() + 1));
        for (int i = 0; i <= word1.size(); i++) dp[i][0] = i;
        for (int j = 0; j <= word2.size(); j++) dp[0][j] = j;
        for (int i = 1; i <= word1.size(); i++) {
            for (int j = 1; j <= word2.size(); j++) {
                if (word1[i - 1] == word2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = std::min({dp[i - 1][j - 1] + 2, dp[i - 1][j] + 1, dp[i][j - 1] + 1});
                }
            }
        }
        return dp[word1.size()][word2.size()];
    }
};

// -------------------------------
int main(int argc, char** argv)
{
    std::string word1 = "sea";
    std::string word2 = "eat";

    Solution solution;
    auto min_steps = solution.minDistance(word1, word2);
    std::cout << "The solution of this problem is " << min_steps << std::endl;
    
    return 0;
}
